distribution/lib/Standard/Database/0.0.0-dev/src/Internal/JDBC_Connection.enso

```diff
@@ -120,31 +118,61 @@ type JDBC_Connection
 
     ## PRIVATE
-
-       Given an insert query template and the associated Database_Table, and a
-       Materialized_Table of data, load to the database.
-    load_table : Text -> Statement_Setter -> Materialized_Table -> Integer -> Nothing
-    load_table self insert_template statement_setter table batch_size =
+       Runs the following action with auto-commit turned off on this connection
+       and returns the result.
+
+       Afterwards, the auto-commit setting is reverted to the state before
+       running this function (so if it was off before, this method may not
+       change anything).
+    run_without_autocommit : Any -> Any
+    run_without_autocommit self ~action =
         self.with_connection java_connection->
             default_autocommit = java_connection.getAutoCommit
-            java_connection.setAutoCommit False
-            Managed_Resource.bracket Nothing (_ -> java_connection.setAutoCommit default_autocommit) _->
-                Managed_Resource.bracket (java_connection.prepareStatement insert_template) .close stmt->
-                    num_rows = table.row_count
-                    columns = table.columns
-                    check_rows updates_array expected_size =
-                        updates = Vector.from_polyglot_array updates_array
-                        if updates.length != expected_size then Panic.throw <| Illegal_State.Error "The batch update unexpectedly affected "+updates.length.to_text+" rows instead of "+expected_size.to_text+"." else
-                            updates.each affected_rows->
-                                if affected_rows != 1 then
-                                    Panic.throw <| Illegal_State.Error "A single update within the batch unexpectedly affected "+affected_rows.to_text+" rows."
-                    0.up_to num_rows . each row_id->
-                        values = columns.map col-> col.at row_id
-                        set_statement_values stmt statement_setter values
-                        stmt.addBatch
-                        if (row_id+1 % batch_size) == 0 then check_rows stmt.executeBatch batch_size
-                    if (num_rows % batch_size) != 0 then check_rows stmt.executeBatch (num_rows % batch_size)
+            Managed_Resource.bracket (java_connection.setAutoCommit False) (_ -> java_connection.setAutoCommit default_autocommit) _->
+                action
+
+    ## PRIVATE
+       Performs the given action within a transaction.
+       Once the action is completed, the transaction is committed.
+       If a panic escapes from the action, the transaction is rolled-back and
+       closed.
+    run_within_transaction : Any -> Any
+    run_within_transaction self ~action =
+        self.run_without_autocommit <|
+            self.with_connection java_connection->
+                handle_panic caught_panic =
+                    java_connection.rollback
```

Comment:
What happens if the rollback throws another Panic?

Reply:
Would it be possible to catch it, and combine the messages into one?

Reply:
Yes, agreed, we should have a ticket for it. It's useful in general (and in this case, when you want to know if the rollback failed), but doesn't seem urgent since it's very unlikely to fail in practice.

Reply:
Added a ticket #6440